### ISSUE TITLE
[Very low priority] Additional options in backup settings

### ISSUE THREAD

**Issue by Cyberavater** [MuntashirAkon/AppManager#237]
1. Create a backup of AM itself as APK in Backup Folder.
As it suggests, enabling it, AM would backup AM as an apk when the user selects AM itself in batch (or single-ly) , Should also contain AM settings and preference there. On the first launch after installation of AM, AM would show a prompt if the user wants to imports the settings from the backup folder. (Somewhat useful after migration);

2. Smart Keystore
Enabling it, AM would check if the device is suitable for Keystore apps to restore, if yes, AM would consider it as a regular app while backing up; if not while backing up, AM should only backup apk of the apps, regardless of what component is selected as other components are useless backing up anyway. While Batch (or single) restoring it'd simply install the apk (As/Till there is any known working method to hack Keystore) (Should help to save some space while batch backup by not collecting useless data).

**Comment by MuntashirAkon** [comment]
> 1. On the first launch after installation of AM, AM would show a prompt if the user wants to imports the settings from the backup folder.

Yeah, seems to be a nice idea. Also, AM currently doesn't handle the restore of AM itself properly (as it force stops the app).

> 2. if not while backing up, AM should only backup apk of the apps, regardless of what component is selected as other components are useless backing up anyway.

Most apps with KeyStore still work fine after a restore. Only a few apps are problematic. If you have further ideas, post them in the relevant threads (#139 or #82) instead.

**Comment by Cyberavater** [comment]
Okay, I've updated the OP accordingly.

**Comment by Atrate** [comment]
Also, AM should not use a constant variable to identify itself. If someone rebuilds it (either through the upcoming apk editing feature or through making a fork) and changes the packagename from `io.github.muntashirakon.AppManager` to something else, it could cause issues.

That's why AM should decide what package it actually is dynamically, through some call.

**Comment by MuntashirAkon** [comment]
> That's why AM should decide what package it actually is dynamically, through some call.

AM already handles this properly as debug and release builds have different signatures.

**Comment by MuntashirAkon** [comment]
The trouble here is the verification of the backups. During the migration and at first launch, the user isn't expected to install OpenKeychain or configure encryption. So, there should be another but easier way to verify the backups of AM. AM stores sensitive information in the shared preferences which can be utilised by third-party apps to call any system services or execute arbitrary commands. Considering the steady rise of its popularity in only 8 months despite no real advertising and a complete absence from Play Store, malware creators would be happy to exploit anything they could find to get into the devices of the supposedly “geeks”.

**Comment by MuntashirAkon** [comment]
With the introduction of App Manager's own KeyStore (#116), it's now possible to verify App Manager's backups without the previously described issues.

**Backup**
1. A custom AES key will be used for encrypting the backups which will be stored in the KeyStore
2. The KeyStore will be stored in the backup path without encryption (the KeyStore already has its own password)
3. The rests are the same as the other backups (with data/external data options may be available for non-root users)

**Restore**
1. Import the KeyStore first (and the ask user for the KeyStore password)
2. Verify backups using the special AES key stored in the KeyStore
3. Restore backups just like any other apps
4. Ask the user to restart the app

**Triggers**
1. When user manually restores backup like any other apps
2. On first launch
3. When user changes the backup location

**Comment by Cyberavater** [comment]
Can you implement an encryption-less backup system for time being?

Proposed model:
- AM'apk and A file with necessary information about AM settings will be in the AppManager folder (could also use a subfolder in here, name starting with 0 or something so that the folder is on top of all folders)
- Upon the first launch after the new installation AM would look for the folder in all possible volumes it can find, if it finds the file, it would ask the user if he would want to load the settings (if found multiple, ie, from all volumes it would ask from which volume the user wants to restore settings from).


I think this issue should be prioritized, as AM's debug build tests(/testers) need to uninstall and install almost every time (Only a few times I was able to get away without uninstalling and configuring my preferred settings every time is really annoying).

**Comment by MuntashirAkon** [comment]
> Can you implement an encryption-less backup system for time being?

No. Because that would introduce compatibility issues in the future. Designs have to be consistent because adding compatibility is a huge redundancy, and maintaining compatibility is also a very tiresome work on developers' side.

> I think this issue should be prioritized, as AM's debug build tests(/testers) need to uninstall and install almost every time (Only a few times I was able to get away without uninstalling and configuring my preferred settings every time is really annoying).

Most of the time, a simple restart should fix the underlying issues. Reinstalling is needed only in a few cases such as room database being incompatible due to version downgrade.

This is a low priority issue because designing a completely new feature takes a lot of time which I currently do not have. Besides, there are more important issues that need attention such as storage issue (#470, #473) and permission issue (#456, #388, #542, #522).

**Comment by MuntashirAkon** [comment]
This is what I've come up with after an initial investigation:

1. App Manager backup should be generic so that another version of App Manager can restore the backup. For example, **AM Debug** should restore **App Manager** or vice versa. This means we cannot use a custom AES key. Instead the user would have to configure the encryption setting first so that it can recover the backup. This is okay because we shall be enforcing backup encryption very soon (See the vulnerability mentioned in #773).
2. The APK of App Manager cannot be stored in plain text in anywhere (unlike what most other backup applications do). This is a potential security risk since an attacker can simply replace it with a compromised version of the app and take control of the entire device once the user installs it. This is so simple to exploit that you don't need to be a very good hacker or work in the NSO group. I cannot put such trapdoor in my app.
3. The following options are considered:
   1. On first launch: App Manager would check for backups in all places it has access to and then ask the user if he wants to restore a backup (and proceed thereby).
   2. Typical restore operation: Restore backups as usual but skip restoring the APK file even if the user has explicitly requested it. As pointed out earlier, we shall still allow the user to backup the APK files like in a regular backup so that the app can be restore by another version of App Manager.
